src/audio/dai-zephyr.c

```diff
@@ -1283,17 +1340,42 @@ int dai_zephyr_copy(struct dai_data *dd, struct comp_dev *dev)
 
 	buffer_release(buf_c);
 
-	buf_c = buffer_acquire(dd->local_buffer);
 
 	/* calculate minimum size to copy */
 	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		buf_c = buffer_acquire(dd->local_buffer);
 		src_samples = audio_stream_get_avail_samples(&buf_c->stream);
 		sink_samples = free_bytes / sampling;
 		samples = MIN(src_samples, sink_samples);
+		buffer_release(buf_c);
 	} else {
+		struct list_item *sink_list;
+
 		src_samples = avail_bytes / sampling;
-		sink_samples = audio_stream_get_free_samples(&buf_c->stream);
-		samples = MIN(src_samples, sink_samples);
+		list_for_item(sink_list, &dev->bsink_list) {
+			struct comp_dev *sink_dev;
+			struct comp_buffer *sink;
+
+			sink = container_of(sink_list, struct comp_buffer, source_list);
+			buf_c = buffer_acquire(sink);
+			sink_dev = buf_c->sink;
+
+			/* this happens when the DAI device is an endpoint created by the copier */
+			if (!sink_dev) {
+				sink_samples = audio_stream_get_free_samples(&buf_c->stream);
```

Comment:
can combine 1365-1366 with 1372-1373?

Reply:
I dont think so, there's no easy way to combine them. This is easier to read I think